### PR TITLE
Rework spidercoordinator informer to update pod and service cidr

### PR DESCRIPTION
**Pull request by cyclinder**
The responsibility of the SpiderCoordinator Informer is to monitor changes in the Pod and Service CIDR of the cluster and automatically update them in the Status of SpiderCoordinator. However, due to different sources of Pod and Service CIDR in different cluster environments, a uniform approach cannot be adopted. Additionally, to avoid coroutine leaks and confusion in updating status,

#### What type of PR is this?

- release/bug 
- release/feature

**What this PR does / why we need it**:

This PR addresses the following four main issues:

1. Avoiding goroutine leakage and avoiding frequent goroutine creation and closure affecting performance. In the previous version, every time the spidercoordiantor's instance is updated, it will close and restart the goroutine that listens to the pod and the service, which may affect the performance or even leak the goroutine in large-scale case.

2. Unify the updating of spidercoordinator status in each reconcile event of spidercoordinator, instead of updating it in different places. Avoid confusing or even unsynchronized status updates.

3. Avoid unnecessary reconcile events, Reduce requests to interact with the APISever.

4. Add a "Reason" field to SpiderCoordinator's status, which can explain why the Phase is `NotReady`.

```
apiVersion: spiderpool.spidernet.io/v2beta1
kind: SpiderCoordinator
metadata:
  creationTimestamp: "2024-02-27T10:18:43Z"
  finalizers:
  - spiderpool.spidernet.io
  generation: 5
  name: default
  resourceVersion: "174300"
  uid: 95891abd-2ab3-4fcf-b5cc-68570a551ee7
spec:
  detectGateway: false
  detectIPConflict: false
  hijackCIDR:
  - 10.244.64.0/18
  - fd00:10:244::/112
  hostRPFilter: 0
  hostRuleTable: 500
  mode: auto
  podCIDRType: calico
  podDefaultRouteNIC: ""
  podMACPrefix: ""
  tunePodRoutes: true
  txQueueLen: 0
status:
  overlayPodCIDR: []
  phase: NotReady
  reason: 'failed to get API group resources: unable to retrieve the complete list
    of server APIs: crd.projectcalico.org/v1: the server could not find the requested
    resource'
  serviceCIDR:
  - 10.233.0.0/18
  - fd00:10:233::/116
```

```
{"level":"INFO","ts":"2024-02-28T07:06:31.313Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:312","msg":"PodCIDRtype changed from calico to cilium","CoordinatorName":"default","Operation":"UPDATE"}
{"level":"DEBUG","ts":"2024-02-28T07:06:31.313Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:314","msg":"Enqueue Coordinator","CoordinatorName":"default","Operation":"UPDATE"}
{"level":"INFO","ts":"2024-02-28T07:06:31.313Z","logger":"spiderpool","caller":"record/event.go:364","msg":"Event(v1.ObjectReference{Kind:\"SpiderCoordinator\", Namespace:\"\", Name:\"default\", UID:\"95891abd-2ab3-4fcf-b5cc-68570a551ee7\", APIVersion:\"spiderpool.spidernet.io/v2beta1\", ResourceVersion:\"180876\", FieldPath:\"\"}): type: 'Normal' reason: 'PodCIDRTypeChanged' Pod CIDR type changed from calico to cilium"}
{"level":"INFO","ts":"2024-02-28T07:06:31.316Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:462","msg":"Trying to fetch the ClusterCIDR from kube-system/kubeadm-config","Event Key":"SpiderCoordinator/default","Operation":"PROCESS"}
{"level":"INFO","ts":"2024-02-28T07:06:31.316Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:483","msg":"Detect podCIDRType is: cilium, try to update podCIDR","Event Key":"SpiderCoordinator/default","Operation":"PROCESS"}
{"level":"WARN","ts":"2024-02-28T07:06:31.319Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:507","msg":"ServiceCIDR is unavailable, update service cidr from cluster service cidr","Event Key":"SpiderCoordinator/default","Operation":"PROCESS"}
{"level":"INFO","ts":"2024-02-28T07:06:31.319Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:435","msg":"Patching coordinator's status from {NotReady failed to get API group resources: unable to retrieve the complete list of server APIs: crd.projectcalico.org/v1: the server could not find the requested resource [] [10.233.0.0/18 fd00:10:233::/116]} to {Synced  [10.244.64.0/18 fd00:10:244::/112] [10.233.0.0/18 fd00:10:233::/116]}","Event Key":"SpiderCoordinator/default","Operation":"PROCESS"}
{"level":"INFO","ts":"2024-02-28T07:06:31.319Z","logger":"spiderpool","caller":"record/event.go:364","msg":"Event(v1.ObjectReference{Kind:\"SpiderCoordinator\", Namespace:\"\", Name:\"default\", UID:\"95891abd-2ab3-4fcf-b5cc-68570a551ee7\", APIVersion:\"spiderpool.spidernet.io/v2beta1\", ResourceVersion:\"180876\", FieldPath:\"\"}): type: 'Warning' reason: 'NetworkingV1alpha1NotFound' ServiceCIDR is available in k8s v1.29, Your cluster version maybe less than v1.29"}
{"level":"INFO","ts":"2024-02-28T07:06:31.357Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:325","msg":"status have no changes, ignore add it to workqueue","CoordinatorName":"default","Operation":"UPDATE"}
{"level":"INFO","ts":"2024-02-28T07:06:31.357Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:440","msg":"Success to patch coordinator's status to {Synced  [10.244.64.0/18 fd00:10:244::/112] [10.233.0.0/18 fd00:10:233::/116]}","Event Key":"SpiderCoordinator/default","Operation":"PROCESS"}
{"level":"INFO","ts":"2024-02-28T07:06:31.357Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:419","msg":"Succeed to SYNC","Event Key":"SpiderCoordinator/default","Operation":"PROCESS"}
```

**Which issue(s) this PR fixes**:

Fixes https://github.com/spidernet-io/spiderpool/issues/3225
Fixes https://github.com/spidernet-io/spiderpool/issues/3210

**Special notes for your reviewer**:

### CODE REVIEW

File: charts/spiderpool/templates/deployment.yaml [hunk 1]
```diff
@@ -175,6 +175,8 @@ spec:
           value: {{ .Values.global.cniConfHostPath | quote }}
         - name: SPIDERPOOL_COORDINATOR_ENABLED
           value: {{ .Values.coordinator.enabled | quote }}
+        - name: SPIDERPOOL_COORDINATOR_DEAFULT_NAME
```
Comment (weizhoublue) [inline charts/spiderpool/templates/deployment.yaml:178]: new environment should be added to document

Comment (cyclinder) [reply charts/spiderpool/templates/deployment.yaml:178]: Added.

File: pkg/k8s/apis/spiderpool.spidernet.io/v2beta1/spidercoordinator_types.go [hunk 2]
```diff
@@ -60,11 +60,14 @@ type CoordinatorStatus struct {
 	// +kubebuilder:validation:Requred
 	Phase string `json:"phase"`
 
+	// +kubebuilder: validation:Optional
+	Reason string `json:"reason,omitempty"`
```
Comment (Icarus9913) [inline pkg/k8s/apis/spiderpool.spidernet.io/v2beta1/spidercoordinator_types.go:64]: *string

Comment (cyclinder) [reply pkg/k8s/apis/spiderpool.spidernet.io/v2beta1/spidercoordinator_types.go:64]: 这里我的想法: 一个 omitempty 足够满足需求，这里不需要考虑默认值和赋空值的区别，这里的需求是reason 为空，不显示，不为空就显示，比较简单。如果为指针类型，还需要处理一些 nil 的情况，没有必要

Comment (cyclinder) [reply pkg/k8s/apis/spiderpool.spidernet.io/v2beta1/spidercoordinator_types.go:64]: 类似参考:

https://github.com/kubernetes/kubernetes/blob/ecd20dc39a608b93d7f1f085b096e35a62fbb216/staging/src/k8s.io/api/core/v1/types.go#L5898

https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/860daa889e41e494e0527d0d5d22c0cee235f7be/api/v1/sriovnetworknodestate_types.go#L90

**Review by weizhoublue — APPROVED** [review]

**Review by ty-dc — APPROVED** [review]

**Pull request merged**